src/esi/Esi.cc

```diff
@@ -1256,8 +1258,20 @@ ESIContext::parse()
 
         PROF_start(esiParsing);
 
-        while (buffered.getRaw() && !flags.error)
-            parseOneBuffer();
+        try {
+            while (buffered.getRaw() && !flags.error)
+                parseOneBuffer();
+
+        } catch (Esi::ErrorDetail &errMsg) { // FIXME: non-const for c_str()
+            debugs(86, DBG_IMPORTANT, errMsg);
```

Comment:
These are errors the client will be seeing but the admin may be able to fix, or pass on to the web dev to fix. The other calls may need to be made exceptions or raised debug level, but I am leaving that out of scope.

Reply:
> The other calls may need to [...] raised debug level, but I am leaving that out of scope.

You have no reasons to leave out of scope the level of a debugs() statement that this PR adds. The exception reporting levels are currently inconsistent -- some exceptions are reported at level 1 and some at level 5. All that reporting code was added (or at least seriously altered) in this PR. Please make that new reporting consistent (or document why it should be different).

Reply:
So you would rather have all these errors silenced for the admin when they should be highlighted for fixing ASAP?

What I am leaving out of scope is the expansion of the *other* debugs which are wrongly quiet today.

Reply:
> So you would rather have all these errors silenced for the admin when they should be highlighted for fixing ASAP?

I would rather have a consistent approach to error reporting rather than inconsistent new debugs() calls justified by ever-changing hand-waving. In the context of level-0/1 debugging, I am using the corresponding rule of thumb:

Admin should only be informed about important events that admins have control over such as (suspected) serious Squid bugs, misconfigurations, attacks, and similar kid/instance-wide failures or state changes. Individual transaction failures (that do not qualify as important events enumerated earlier) should be reflected in access.log records instead. This rule reduces chances that admins stop paying attention to important log messages because the logs are dominated by irrelevant-to-the-admin noise.

Furthermore, important events (as defined above) that may be logged frequently (e.g., those that may be triggered by incoming data) should have a reporting limit. In other words, Squid should stop reporting them after a while or, better, rate-limit such reports. This secondary rule both prevents DoS attacks via excessive logging and makes sure that important messages do not become noise themselves.

```C++
         try {
             ...
         } catch (Esi::ErrorDetail &errMsg) { // FIXME: non-const for c_str()
            // DBG_IMPORTANT because these are syntax bugs the ogirins' dev can fix
            debugs(86, DBG_IMPORTANT, "ERROR: " << errMsg);
            ...
        } catch (...) {
            // level-5 because these are OS issues not easily fixed
```


Now I am applying these rules of thumb to the current PR code quoted above:

* Transaction-specific ESI syntax errors that the admin usually has no direct control over are reported as DBG_IMPORTANT but probably should not be. IIRC, ESI is enabled by default, even on non-accelerating Squids. If that is true, then the fact that "origin developers" can fix the problem is irrelevant -- there is virtually always somebody somewhere who can fix the problem! What is relevant here is that the admin probably cannot fix this problem, and that the problem has transaction scope.

* Unclassified errors, possibly including Squid code bugs detected via Must() failures, are not reported as DBG_IMPORTANT. It is debatable whether all unclassified problems should be reported just because some of them could be important Squid bugs. However, if we do decide to report transaction-specific ESI syntax errors (the first catch clause), then we should be eager to report unclassified errors that may have much wider scope/impact and may be under admin control.

Going forward, I see three ways to address the current problems:

1. Report classified problems at level 2 (or higher) because they are transaction-level problems outside admin control. Report unclassified problems at `DBG_IMPORTANT` because some of them can be Squid bugs. Add a hard-coded limit to `DBG_IMPORTANT`reports.
2. Report both classified and unclassified problems at level 2 (or higher). Let origin admins deal with their ESI problems and assume that all unclassified errors are benign transaction-specific problems. To use this option, we need to convince ourselves that unclassified errors are (usually) unimportant. You may already know enough about ESI code to give examples that will confirm that.
3. Report both classified and unclassified problems at `DBG_IMPORTANT`. Add a hard-coded limit(s) to those reports. To use this option, we need to convince ourselves that the classified errors are important events under admin control. I am not convinced they are.

I suspect option 1 is the best way forward but I am happy to evaluate arguments supporting option 2 and even option 3.

P.S. And if the `ogirins' dev can fix` comment survives these changes, please spellcheck it. And the `OS issues` phrasing would have to be adjusted -- we cannot know what all of those exceptions are and should not pretend otherwise. What "OS issues" did you have in mind?

Reply:
We seem to be disagreeing on the "outside admin control" aspect here. 
* All the client receiving the errors can do is contact this admin.
* The site dev can only see what the client sees. Which is a somewhat obfuscated view of the error.
* The admin is the person with most control in these situations - they have access to the proxy debug, and usually a CDN contract agreement + contact details to pass that info to the dev who can implement the fix.

I agree with option 1, making the protocol WARNING and ERROR level-2 messages to display alongside the protocol I/O and slightly higher than the (level 3+) parse activity. If we agree on that as a general policy for all protocols I will update the *touched* lines here.  I still do not want to bloat this PR with more scope creep just for polishing [this *is* a security related PR after all]. Will do a followup PR that brings the other error/warning/exception handing inline with policy and update the debugs() documentation.

Reply:
I am reordering your response in hope to make progress faster:

> I agree with option 1, 

Great! Please let me know if you want me to implement it to minimize misunderstandings regarding what "option 1" means. I am glad that you liked that option, and would certainly welcome your matching implementation, but I am concerned that your option 1 summary (i.e. `making the protocol WARNING and ERROR level-2 messages`) may not match how I would interpret that option.

> making the protocol WARNING and ERROR level-2 messages

If "protocol messages" are those corresponding to the first `catch` clause in the PR, then (they are what I called "classified problems" in my option1 description, and) we are in agreement here. Level 2 is an acceptable debugging level for them.

However, please note that my option1 also included the second part: Report unclassified problems (i.e. the second `catch` clause) at `DBG_IMPORTANT`. If you are OK with that, than we are in a complete agreement!

> If we agree on that as a general policy for all protocols I will update the _touched_ lines here.

Changing the PR-touched lines is all I am asking for.

> I still do not want to bloat this PR with more scope creep just for polishing

I am not asking for changing any debugs() lines other than those already touched/moved/added by this PR.

Again, I can make those option1 changes if you want me to.

The rest is about future work and further analysis...

----

> Will do a followup PR that brings the other error/warning/exception handing inline with policy and update the debugs() documentation.

Thank you. Before you dive into that development, let's make sure we do agree what those changes should look like. While I think my proposal was rather specific, your seemingly abstract reaction to it does not give me confidence that we are on the same page.

* If you do agree with the two-paragraph rule of thumb that I proposed in my previous [comment](https://github.com/squid-cache/squid/pull/411#discussion_r362064319) and polished below, then perhaps we can start with a single small adjustment or two?
* Otherwise, let's discuss how we can adjust those rules to reach consensus (before making many changes outside this PR?).


----

> We seem to be disagreeing on the "outside admin control" aspect here.

I think the disagreement may be more along the definition of "control". You may be talking "control" as in "can report the problem details to the site admin". I meant "control" as in "views the problem as something he or she should be involved with _and_ can adjust squid.conf or change Squid version to fix the problem". In retrospect, "control over a problem" was a poor expression of the idea.

Here is an adjusted version of the rules of thumb proposed in my previous [comment](https://github.com/squid-cache/squid/pull/411#discussion_r362064319):

When reporting various errors, bugs, and problems:

1. Level-0/1 messages should be used for important events affecting the Squid instance as a whole (e.g., (suspected) serious Squid bugs, misconfigurations, and attacks).
2. Individual transaction failures (that do not qualify as important events in item 1) should be reflected in access.log records.
3. Level-0/1 messages that may be logged frequently (e.g., those that may be triggered by incoming data) should have a reporting limit. In other words, Squid should stop reporting them after a while or, better, rate-limit such reports.

This rule of thumb reduces chances that admins stop paying attention to important cache.log messages because the cache.log is dominated by irrelevant-to-the-admin noise and/or relevant-but-overwhelming info. The rule also prevents DoS attacks via excessive cache.log logging.

N.B. Folks may never agree on whether some particular debugs() line qualifies as item1 problem -- there are too many vague terms in item1 definition. The only solution to _that_ problem (that I know of) is making level-0/1 reporting configurable at individual message level. Factory has a patch for that; we are polishing it for the official submission.

Reply:
I believe option1 is now implemented as of commit ca3a52c. If there are any other changes you were expecting to see please do implement as a sub-branch so we can compare.

Reply:
> I believe option1 is now implemented as of commit ca3a52c.

Thank you. The ball is on my side.

Reply:
I agree that option 1 is now implemented. Resolving this change request.